### PR TITLE
`Development`: Add exam database query count tests

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context

The database query count test is needed to ensure that the Quiz Pool feature does not add additional queries when it is not used, especially in the critical state such as when the students start to work on an exam and also when they submit the exam.

### Description

Add 2 tests that assert the number of database query call when a student starts working on an exam and also when a student submits the exam.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
- [x] Code Review 3

### CODE REVIEW

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java [hunk 1]
```diff
@@ -76,4 +86,78 @@ void testGetAllCoursesForDashboardRealisticQueryCount() throws Exception {
...
+        Exam exam = createActiveExam();
+        StudentExam studentExam = createStudentExamForUser(exam, TEST_PREFIX + "student3");
+        startWorkingOnExam(studentExam);
+
+        assertThatDb(() -> {
+
+            submitExam(studentExam);
+            return null;
+
+        }).hasBeenCalledAtMostTimes(getSubmitAnswerOfExamExpectedTotalQueryCount());
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java:116]: You could shorten this either by changing `submitExam` to return the server response or by adding an overloaded `assertThatDb` method taking a Consumer.
```suggestion
        assertThatDb(() -> submitExam(studentExam))
        .hasBeenCalledAtMostTimes(getSubmitAnswerOfExamExpectedTotalQueryCount());
```

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java [hunk 2]
```diff
@@ -76,4 +86,78 @@ void testGetAllCoursesForDashboardRealisticQueryCount() throws Exception {
...
+    }
+
+    private Exam createActiveExam() {
+        Course course = courseUtilService.addEmptyCourse();
+        ZonedDateTime visibleDate = now().minusMinutes(10);
+        ZonedDateTime startDate = now().minusMinutes(5);
+        ZonedDateTime endDate = now().plusMinutes(5);
+        return examUtilService.addExam(course, visibleDate, startDate, endDate);
+    }
```
Comment (laadvo) [inline src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java:125]: Consider moving this into `ExamUtilService`

**Review by laadvo — COMMENTED** [review]
I agree with @DominikRemo you should try using as little resources as possible and have only one test instead of two

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java (hunk 1 above)
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java]: Changed the submitExam to return the server response which shortened the assertion. Adding an overloaded method with a Consumer argument requires creating a new QueryCountAssertion class as QueryCountAsserttion class requires a ThrowingProducer class.

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java [hunk 3]
```diff
@@ -24,12 +31,15 @@ class DatabaseQueryCountTest extends AbstractSpringIntegrationBambooBitbucketJir
...
+    @Autowired
+    private ExamUtilService examUtilService;
+
     private static final int NUMBER_OF_TUTORS = 1;
 
     @BeforeEach
     void setup() {
         participantScoreScheduleService.shutdown();
-        userUtilService.addUsers(TEST_PREFIX, 1, NUMBER_OF_TUTORS, 0, 0);
+        userUtilService.addUsers(TEST_PREFIX, 3, NUMBER_OF_TUTORS, 0, 0);
```
[the remark this answers is not in the document]
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java]: Modified this to use only 1 user.

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java [hunk 4]
```diff
@@ -76,4 +86,78 @@ void testGetAllCoursesForDashboardRealisticQueryCount() throws Exception {
...
+        StudentExam studentExam = createStudentExamForUser(exam, TEST_PREFIX + "student3");
+        startWorkingOnExam(studentExam);
+
+        assertThatDb(() -> {
+
+            submitExam(studentExam);
+            return null;
+
+        }).hasBeenCalledAtMostTimes(getSubmitAnswerOfExamExpectedTotalQueryCount());
+    }
```
[the remark this answers is not in the document]
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java]: Combined the 2 tests into 1 test.

File: src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java (hunk 2 above)
Comment (rriyaldhi) [reply src/test/java/de/tum/in/www1/artemis/DatabaseQueryCountTest.java]: The method is moved into ExamUtilService.

**Review by Strohgelaender — APPROVED** [review]

**Review by DominikRemo — APPROVED** [review]
Thank you for addressing my comments :)

**Review by laadvo — APPROVED** [review]
looks good now :+1:

**Pull request merged**